Apply pathloss correction to NIRSpec and NIRISS SOSS data

jdavies-st opened the pull request:
- Applies the computed pathloss correction to the `data, err, var_poisson` members of the datamodel
- Applies the `pathloss_pointsource2D` array if `SRCTYPE == "POINT"`, `pathloss_uniformsource2D` if `SRCTYPE == "EXTENDED"` or `None`.
- No longer attaches the various `pathloss_xxx` extensions.  Only a single `pathloss` extension representing the 2D correction applied.

### Questions:

1) ~~For NIRSpec IFU, is the distinction between point source and uniform source meaningful?  I.e. for a point source in the field, some slits will contain it and some won't.  Is this taken care of in the pathloss reffile?~~ This is taken care of in the reffile.

2) ~~It seems like the NIRISS SOSS (non-TSO) version of this step only computes a point source correction.  Should we be checking to make sure SRCTYPE == "POINT" before even computing the correction, and then apply it?~~ Instead of keying off SRCTYPE, this keys off aperture name.  It only gets applied for `SUBSTRIP96` aperture.  More info [here](https://jwst-docs.stsci.edu/display/JTI/NIRISS+Single+Object+Slitless+Spectroscopy#NIRISSSingleObjectSlitlessSpectroscopy-SOSSsubarrays)


Resolves #1152

jdavies-st commented:
It might also be worth removing the `pathloss_xxx` members from the datamodel schemas, if we're decided that we definitely don't want to attach these corrections anymore.  That should be fine with existing data, as if they exist, they will just get shuffled off to `extra_fits`.

hbushouse commented:
> It might also be worth removing the `pathloss_xxx` members from the datamodel schemas, if we're decided that we definitely don't want to attach these corrections anymore. That should be fine with existing data, as if they exist, they will just get shuffled off to `extra_fits`.

See the last comments in https://jira.stsci.edu/browse/JP-240

hbushouse commented on docs/jwst/pathloss/description.rst:
```diff
@@ -36,7 +36,7 @@ It creates 2 pairs of 1-d arrays, a wavelength array (calculated from the WCS ap
 calculated by interpolating each plane of the pathloss cube at the position of
 the source (which is taken from datamodel).  There are pairs of these arrays for
-both pointsource and uniformsource data types.
+both point source and uniform source data types.
```
I think the whole doc needs a review and updates, now that nothing is done with the 1-D corrections. For example, up in line 9 it says this is the correction "to apply to spectra when the 1-d extraction is performed." That's no longer true. It's being applied here and now, at the 2-D stage. And we also need to remove or reword the couple of mentions regarding MIRI corrections not yet being implemented: it's more appropriate to say that correction algorithms for MIRI MRS and LRS modes have not yet been developed (not implemented could be construed as indicating the algorithms exist, but we just haven't gotten around to putting them in yet.)

hbushouse commented on jwst/pathloss/pathloss.py:
```diff
@@ -213,26 +212,31 @@ def do_correction(input_model, pathloss_model):
-                        pathloss_uniformsource_2d = interpolate_onto_grid(wavelength_array,
-                                                                          wavelength_uniformsource,
-                                                                          pathloss_uniform_vector)
-                        slit.pathloss_pointsource2d = pathloss_pointsource_2d
```
Should still save whichever version (point or uniform) of the correction is applied. So this should be moved down into the if-blocks starting at line 224.

hbushouse commented on jwst/pathloss/pathloss.py:
```diff
@@ -270,37 +274,42 @@ def do_correction(input_model, pathloss_model):
-                    pathloss_uniformsource_2d = interpolate_onto_grid(wavelength_array,
-                                                                      wavelength_uniformsource,
-                                                                      pathloss_uniform_vector)
-                    slit.pathloss_pointsource2d = pathloss_pointsource_2d
```
Same comment as above (save the version that gets applied to the slit)

hbushouse commented on jwst/pathloss/pathloss.py:
```diff
@@ -335,18 +340,29 @@ def do_correction(input_model, pathloss_model):
-        pathloss_uniformsource_2d = interpolate_onto_grid(wavelength_array,
-                                                          wavelength_uniformsource,
-                                                          pathloss_uniform_vector)
-        input_model.pathloss_pointsource2d = pathloss_pointsource_2d
```
Save whichever version gets applied.

hbushouse commented on jwst/pathloss/pathloss.py:
```diff
@@ -335,18 +340,29 @@ def do_correction(input_model, pathloss_model):
         # This might be useful to other steps
-        input_model.wavelength = wavelength_array
-        input_model.meta.cal_step.pathloss = 'COMPLETE'
+        output_model.wavelength = wavelength_array
```
Looks like this wavelength array is saved to the output only for NIRSpec IFU mode, correct? I don't think we create a wavelength array for MIRI MRS exposures, so that's inconsistent for the 2 IFU's. @jemorrison is this wavelength array that's in the NIRSpec IFU 2-D slice images ever used anywhere downstream?

jdavies-st replied:
Yeah, I don't think we need to save this wavelength array.  Unless it gets recomputed down the line and would be useful to have.  @jemorrison?  @nden?

hbushouse commented on jwst/datamodels/schemas/pathlosscorr.schema.yaml:
```diff
@@ -5,30 +5,9 @@ properties:
     fits_hdu: PATHLOSS_POINTSOURCE2D
```
I would suggest removing the "2D" suffix, because we no longer need to identify it as such, and perhaps just shorten the whole name (for readability and typability) to just "PATHLOSS_POINT".

jdavies-st replied:
I would suggest we remove the `_point` and `_uniform` from the attached name.  These 2D corrections are of the same type so they should have the same extension name.  The pipeline processing and SRC_TYPE keyword tells which was done.  This means that anything that needs to use these downstream can just read the `pathloss` extension instead of having to figure out whether it should read the `pathloss_point` or `pathloss_uniform` extension.

hbushouse replied:
That's a good point (the potential use downstream needing to know which name to use). I was thinking of keeping points/uniform in the names just as history or logging, but you've convinced me that keeping them generic is probably best.

hbushouse commented on jwst/datamodels/schemas/pathlosscorr.schema.yaml:
```diff
@@ -5,30 +5,9 @@ properties:
-    datatype: float32
   pathloss_uniformsource2d:
     title: 2-d array for pathloss (uniform source)
     fits_hdu: PATHLOSS_UNIFORMSOURCE2D
```
As above, I suggest changing this to just "PATHLOSS_UNIFORM"

hbushouse commented on jwst/pathloss/pathloss.py:
```diff
@@ -213,26 +212,31 @@ def do_correction(input_model, pathloss_model):
+                                pathloss_uniform_vector)
+                            slit.data /= pathloss_uniformsource_2d
+                            slit.err /= pathloss_uniformsource_2d
+                            slit.var_poisson /= pathloss_uniformsource_2d**2
```
Need to save either pathloss_pointsource_2d or pathloss_uniformsource_2d - whichever one was applied - to the slit instance, so it appears in the output product.

hbushouse commented on jwst/pathloss/pathloss.py:
```diff
@@ -270,37 +274,42 @@ def do_correction(input_model, pathloss_model):
+                            pathloss_uniform_vector)
+                        slit.data /= pathloss_uniformsource_2d
+                        slit.err /= pathloss_uniformsource_2d
+                        slit.var_poisson /= pathloss_uniformsource_2d**2
```
Need to save pathloss_pointsource_2d or pathloss_uniformsource_2d to the slit instance.

hbushouse commented on jwst/pathloss/pathloss.py:
```diff
@@ -335,18 +340,29 @@ def do_correction(input_model, pathloss_model):
+            output_model.data /= pathloss_uniformsource_2d
+            output_model.err /= pathloss_uniformsource_2d
+            output_model.var_poisson /= pathloss_uniformsource_2d**2
+
```
Save to the output model whichever correction array gets applied.

jdavies-st commented:
I've fixed up everything from the comments.  I still need input on how to apply the correction for the NIRISS SOSS non-TSO data.  As far as I can tell, the correction is a 1D array that is not the same shape as the input data model.  Do we have any example data that could point the way forward?

Or I can figure that out in a separate PR.

hbushouse commented on jwst/datamodels/schemas/pathlosscorr.schema.yaml:
```diff
@@ -1,34 +1,8 @@
-    title: 2-d array for pathloss (point source)
-    fits_hdu: PATHLOSS_POINTSOURCE2D
+  pathloss:
+    title: 2D array for pathloss
```
Minor nit, but for consistency with title of barshadow extension, which is "Bar shadow correction", why don't we make this "Path loss correction".

jdavies-st commented:
Added regression test for `PathLossStep` using NIRISS SOSS data and updated the NIRSPEC IFU regression test to not go looking for `pathloss_foo` extensions.  Regression test truth data will still need to be updated after this is merged.

Also fixed the base class `BaseJWSTTestSteps` so it doesn't overwrite its inputs, which addresses some of #2941.  That bug affected the new regression test.

hbushouse commented:
I know this is already closed, but I just noticed that when the pathloss extension was added back to the ``ImageModel`` schema (for NIRISS SOSS data) it was given a default value of 1.0. This may cause that extension to show up in all ``ImageModel``s that get saved to disk, even for imaging modes that don't need or use it. That should be checked. Removing the line ``default: 1.0`` may prevent it from showing up when it isn't needed.

jdavies-st commented:
The `pathlosscorr.schema` is what got added to `image.schema`, not the `pathloss.schema`.  The later is for the pathloss reference file.  The `.pathloss` attribute attached when the step is successfully run has no default value.

hbushouse commented:
Sorry, I should've been more specific in my comment. Line 53 of ``image.schema.yaml`` https://github.com/spacetelescope/jwst/pull/3323/files#diff-e3b3f74a5303d3dea3d6067ee01ca53dR53 added a reference to include ``pathloss.schema.yaml``, which defines a 2-d array with an attribute name of ``pathloss`` and fits_hdu of ``PATHLOSS``.

However, line 34 of ``image.schema.yaml`` https://github.com/spacetelescope/jwst/blob/master/jwst/datamodels/schemas/image.schema.yaml#L34 already contained the definition of 2-d data array with attribute name ``pathloss_pointsource`` and fits_hdu of ``PATHLOSS_POINTSOURCE``. So now we've got 2 pathloss correction arrays defined in ``ImageModel``. It's this second one, ``pathloss_pointsource``, that has default=1.0 set for it, which I believe will cause it to appear in all output files regardless of whether or not we actually populate it in the code.

So at the very least we need to clean up the double definition of these 2 pathloss-related arrays.

jdavies-st commented:
It doesn't get created because there's no auto-access in the datamodel of `self.pathloss_pointsource` as there is for `self.data`, etc.  But it should be removed.  Will do so.  Thanks for the 🦅 👁 .
